ENH Adds missing value support to OneHotEncoder

thomasjpfan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes https://github.com/scikit-learn/scikit-learn/issues/11996
Closes https://github.com/scikit-learn/scikit-learn/pull/15009
Closes https://github.com/scikit-learn/scikit-learn/pull/13028
Towards https://github.com/scikit-learn/scikit-learn/issues/15796


#### What does this implement/fix? Explain your changes.
Adds missing value support to `OneHotEncoder`.

For numerical data, `np.nan` is represents missing values. For object dtypes, `None` and `np.nan` is support for missing values.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thomasjpfan committed:
ENH Adds support for missing values

thomasjpfan committed:
BUG Fix

thomasjpfan committed:
DOC More docstrings

thomasjpfan committed:
DOC Adds comments

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -350,6 +363,27 @@ def _compute_drop_idx(self):
+            drop_idxs = []
+            for i, (val, cat_list) in enumerate(zip(self.drop,
+                                                    self.categories_)):
+                if not is_scalar_nan(val):
```
So this is for the user explicitly specifying that nan should be dropped? Because 'first' will never result in nan being dropped, right?

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -350,6 +363,27 @@ def _compute_drop_idx(self):
             missing_drops = [(i, val) for i, val in enumerate(self.drop)
                              if val not in self.categories_[i]]
+
+            missing_drops = []
```
Doesn't the missing_drops implementation from two lines above still work? And if not you should at least remove it? but set operations should work on NaN, right?

amueller replied:
this hasn't been addressed, right? There is still an implementation two lines above that is then overwritten. And I am pretty sure that the old code still works. you can also use list.index which works on NaN as expected.

thomasjpfan replied:
Since `self.categories_[i]` is a numpy array, I was trying to avoid converting it into another data structure and making a copy.

amueller commented on sklearn/utils/_encode.py:
```diff
@@ -27,13 +28,47 @@ def _unique(values, *, return_inverse=False):
+        nan_in_set = np.nan in uniques_set
+
+        if none_in_set and nan_in_set:
+            raise ValueError("Input wiith both types of missing, None and "
```
```suggestion
            raise ValueError("Input with both types of missing, None and "
```

amueller replied:
Why is this not supported? Because the order would be unclear?

amueller commented on sklearn/utils/_encode.py:
```diff
@@ -118,26 +153,57 @@ def _check_unknown(values, known_values, return_mask=False):
+        nan_in_diff = np.nan in diff
+
+        if none_in_diff and nan_in_diff:
+            raise ValueError("Input wiith both types of missing, None and "
```
```suggestion
            raise ValueError("Input with both types of missing, None and "
```

amueller commented on sklearn/utils/_encode.py:
```diff
@@ -118,26 +153,57 @@ def _check_unknown(values, known_values, return_mask=False):
+        if none_in_diff and nan_in_diff:
+            raise ValueError("Input wiith both types of missing, None and "
+                             "np.nan, is not supported")
+        if none_in_diff:
```
Why is this necessary? Does ``sort`` fail to sort? Can you add a comment?

thomasjpfan replied:
This comment may have lost its context. The `nans` were removed from the sets above because they do not work with set differencing in this context.

As for the `Nones`, we could have left the `Nones` in the set, but it would complicate the `_extract_missing` helper.

amueller commented on sklearn/utils/_encode.py:
```diff
@@ -118,26 +153,57 @@ def _check_unknown(values, known_values, return_mask=False):
+
+    if return_mask:
+        return diff, valid_mask
+    # valid_mask is
```
?

amueller commented on sklearn/utils/tests/test_encode.py:
```diff
@@ -53,21 +62,117 @@ def test_encode_with_check_unknown():
+def test_unique_util_missing_values_error_both_missing():
+    # _unique does not support both types of missing
+    values = np.array(['a', 'c', 'c', None, np.nan], dtype=object)
+    msg = ("Input wiith both types of missing, None and np.nan, is not "
```
```suggestion
    msg = ("Input with both types of missing, None and np.nan, is not "
```

amueller commented on sklearn/utils/tests/test_encode.py:
```diff
@@ -53,21 +62,117 @@ def test_encode_with_check_unknown():
+def test_check_unknown_errors_both_missing_values():
+    # _check_unknown does not support both types of missing
+    values = np.array(['a', 'c', 'c', None, np.nan], dtype=object)
+    msg = ("Input wiith both types of missing, None and np.nan, is not "
```
```suggestion
    msg = ("Input with both types of missing, None and np.nan, is not "
```

amueller reviewed (commented):
Looks good but needs docs and a whatsnew. Also, you should explain why using None and NaN in a single column is prohibited.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
DOC Adds to user guide

thomasjpfan committed:
ENH Remove imputer from example

thomasjpfan committed:
ENH Support for both nan and None

thomasjpfan committed:
CLN Remove unneeded comment

thomasjpfan commented:
Updated PR with:

1. Support for both `np.nan` and `None` support for object dtype.
2. Updated examples that do not need an imputer anymore before the `OneHotEncoder`.

NicolasHug commented:
What's the motivation for treating None as missing?

NicolasHug commented on doc/modules/preprocessing.rst:
```diff
@@ -590,6 +590,22 @@ In the transformed `X`, the first column is the encoding of the feature with
 the 2 features with respectively 3 categories each.
 
+:class:`OneHotEncoder` supports categorical features with missing values by
+considering themissing values its own category::
```
```suggestion
considering the missing values as a category::
```

NicolasHug commented on doc/whats_new/v0.24.rst:
```diff
@@ -129,6 +129,9 @@ Changelog
 ............................
 
+- |Feature| :class:`preprocessing.OneHotEncoder` now supports missing
+  values by treating them as its own category. :pr:`17317` by `Thomas Fan`_.
```
```suggestion
  values by treating them as a category. :pr:`17317` by `Thomas Fan`_.
```

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -87,9 +90,17 @@ def _fit(self, X, handle_unknown='error'):
+                                 "supported for numerical categories")
+                    # np.nan should be the last element
+                    if np.isnan(sorted_cats[-1]):
+                        if not (np.isnan(cats[-1]) or
+                                np.all(sorted_cats[:-1] == cats[-1])):
```
should this be

```py
                        if not (np.isnan(cats[-1]) and
                                np.all(sorted_cats[:-1] == cats[:-1])):
```

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -87,9 +90,17 @@ def _fit(self, X, handle_unknown='error'):
+                    sorted_cats = np.sort(cats)
+                    error_msg = ("Unsorted categories are not "
+                                 "supported for numerical categories")
+                    # np.nan should be the last element
```
```suggestion
                    # if there are nans, np.nan should be the last element
```

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -98,8 +109,10 @@ def _fit(self, X, handle_unknown='error'):
 
-    def _transform(self, X, handle_unknown='error'):
-        X_list, n_samples, n_features = self._check_X(X)
+    def _transform(self, X, handle_unknown='error',
+                   force_all_finite=False):
```
doesn't this it in one line?

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
CLN Apply suggestions

thomasjpfan committed:
ENH Apply suggestions

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

jnothman commented:
Test failures, @thomasjpfan

thomasjpfan committed:
BUG Fix

thomasjpfan commented:
Test failures are fixed!

amueller commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -350,6 +362,27 @@ def _compute_drop_idx(self):
+                    continue
+
+                # val is nan, find nan in categories manually
+                for i, cat in enumerate(cat_list):
```
you're overloading i in a confusing way.

amueller commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -535,24 +505,6 @@ def test_ordinal_encoder_inverse():
-@pytest.mark.parametrize("X", [np.array([[1, np.nan]]).T,
-                               np.array([['a', np.nan]], dtype=object).T],
-                         ids=['numeric', 'object'])
-def test_ordinal_encoder_raise_missing(X):
```
What's the behavior for ordinal encoder?

amueller commented on sklearn/utils/_encode.py:
```diff
@@ -27,13 +28,36 @@ def _unique(values, *, return_inverse=False):
     try:
-        uniques = sorted(set(values))
+        uniques_set = set(values)
+        missing_values_in_set = [value for value in (None, np.nan)
```
is None not sortable?

jnothman replied:
This won't match `float('nan')`, nor `(np.array(0)/0).item()`. Does it matter?

To be careful, I think you might be safer with:
```py
        missing_values_in_set = [value for value in uniques_set
                                 if value is None or is_scalar_nan(value)]
```

amueller commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -690,9 +642,229 @@ def test_encoders_has_categorical_tags(Encoder):
+
+
+def test_ohe_missing_value_object_different_order():
+    # categories are not in lexicon ordering
```
```suggestion
    # categories are not in lexical ordering
```

amueller commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -690,9 +642,229 @@ def test_encoders_has_categorical_tags(Encoder):
+def test_ohe_missing_values_both_missing_values():
+    # test both types of missing of missing values are treated as its own
+    # category
+    X = np.array([['a', 'b', None, 'a', np.nan]], dtype=object).T
```
not sure if it's necessary but having ``np.nan`` twice would check the deduplication logic (again, I know it's covered in other tests as well).

amueller reviewed (commented):
Looks good. I am still slightly hopeful two places I comment on can be simplified, otherwise ready to merge I'd say :)

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
ENH Fix bug

thomasjpfan committed:
ENH Rename variables to be more clear

thomasjpfan committed:
CLN Adds more comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

justmarkham commented:
@thomasjpfan This is an exciting feature - thanks for your work on this!

I have one question, if you don't mind me asking: What will happen if a missing value is not seen during fit, but it is seen during transform? Will the OneHotEncoder throw an error?

(Sorry to ask, but I'm not fluent enough with reading scikit-learn code to figure this out on my own.)

Thanks!

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -87,9 +90,17 @@ def _fit(self, X, handle_unknown='error'):
+                    error_msg = ("Unsorted categories are not "
+                                 "supported for numerical categories")
+                    # if there are nans, nan should be the last element
+                    if np.isnan(sorted_cats[-1]):
+                        if not (np.isnan(cats[-1]) or
+                                np.all(sorted_cats[:-1] == cats[:-1])):
+                            raise ValueError(error_msg)
+                    else:
+                        if not np.all(sorted_cats == cats):
+                            raise ValueError(error_msg)
```
I find this logic hard to read. Is the following better?
```suggestion
                    stop_idx = -1 if np.isnan(sorted_cats[-1]) else None
                    if (not np.all(sorted_cats[:stop_idx] == cats[:stop_idx]) or
                            (np.isnan(sorted_cats[-1]) and not np.isnan(cats[-1]))):
                        raise ValueError(error_msg)
```

jnothman replied:
I don't think the `np.isnan(sorted_cats[-1]) and not np.isnan(cats[-1])` is tested

jnothman commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -690,9 +660,229 @@ def test_encoders_has_categorical_tags(Encoder):
+])
+def test_ohe_missing_value_mixed_sanity(categories):
+    # sanity check for mixed data with missing values
+    X = np.array([['a', 'b', None], [0, np.nan, 2]], dtype=object).T
```
I find the use of `.T` with multiple samples and features easy to miss... maybe use .transpose() instead, or avoid it.

jnothman commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -690,9 +660,229 @@ def test_encoders_has_categorical_tags(Encoder):
+
+
+def test_ohe_missing_values_both_missing_values():
+    # test both types of missing of missing values are treated as its own
+    # category
```
```suggestion
    # test both types of missing values are treated as its own category
```

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -27,13 +28,36 @@ def _unique(values, *, return_inverse=False):
+        missing_values_in_set = [value for value in (None, np.nan)
+                                 if value in uniques_set]
+
+        uniques_set -= set(missing_values_in_set)
```
Just to be careful:
```py
>>> {np.nan} - {np.nan}
set()
>>> {float('nan')} - {np.nan}
{nan}
```

Could this be an issue?

jnothman reviewed (commented):
What is the current behaviour with pd.NA? I think it should be tested, whether or not we support it.

I don't know if there are practical use-cases where it matters, but you've tried to rely on python object identity (`is`) to identify NaNs in `_encode.py`, comparing them to `np.nan`. This won't work for NaNs obtained by other (nefarious!) means. You need to strictly use `np.isnan` or `is_scalar_nan`.

[I'm also a bit surprised that the OHE tests are not able to reuse a bit more of existing tests, seeing as we're essentially parametrising the value of the highest category... but I know nans can be tricky. It would be nice if more existing tests were reused. This might also help with testing more parameter combinations (in the future, for instance). But you can tell me to bugger off.]

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan commented:
This is okay with the other formats since `np.array` will convert `float('nan')` into a `np.nan`. This is the same for `list` and `check_array`:

```py
X = [[float('nan'), 1], [0, 0], [float('nan'), 1]]
X_checked = check_array(X, force_all_finite='allow-nan')
type(X_checked[0, 0])
# numpy.float64
```

This means that `float('nan')` and `np.nan` are treated the same when the dtype is float:

```py
X = np.array([float('nan'), np.nan])
type(X[0]), type(X[1])
# (numpy.float64, numpy.float64)
```

The cases where `float('nan')` becomes an issue is with a ndarray with object dtype

```py
X = np.array([[float('nan'), 'd'], ['b', 'a'], [float('nan'), 'a']], dtype=object)

# or

df = pd.DataFrame({'a':  ['b', float('nan')]})
```

For this PR, I purposefully did not handle these situation. Have you see data encoded like this before? 

If we were to handle this situation, should we treat `float('nan')` and `np.nan` the same since they are treated the same for `dtype=float`?

jnothman commented:
> If we were to handle this situation, should we treat float('nan') and np.nan the same since they are treated the same for dtype=float?

I think behaviour would ordinarily be unexpected if different ways to get NaN were not handled the same, although I admit that in the context of categoricals, `0/0` is a strange way to acquire NaN (unless one was not aware of np.NaN).

I think we should be good citizens and do one of:
* document this caveat, possibly erroring when we identify that one of the cateogries is a non-NumPy NaN; or
* test and handle alternative NaNs.

What is the runtime cost of changing:
```py
missing_values_in_set = [value for value in (None, np.nan) if value in uniques_set]
```
to
```py
missing_values_in_set = [value for value in uniques_set if value is None or is_scalar_nan(value)]
```
... and if much slower as a proportion of encoding, can we rely on `pd.isna` when pandas is installed for speed.


I wouldn't expect this is a bottleneck, seeing as we're working over uniques.

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
ENH Raise when float('nan') is in encoder

thomasjpfan committed:
DOC Apply suggestions

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
TST Adds sorted test

thomasjpfan committed:
TST Combines more tests

thomasjpfan committed:
TST Combines more tests

thomasjpfan committed:
ENH Removes unneeded function

thomasjpfan committed:
TST Fixes test

thomasjpfan committed:
TST Fixes test

thomasjpfan committed:
DOC Adds note to user guide

thomasjpfan committed:
ENH Error on float('nan') for custom catgories

thomasjpfan committed:
TST Fix

thomasjpfan commented:
@justmarkham 

> Since this current PR causes missing values to be treated as a new category, a missing value seen during transform will also be encoded as all zeros (and thus will not throw an error). Is that correct?

That is correct

@jnothman Updated with:

1. Errors when `float('nan')` is used. We can add this support in the future if there is demand for it.
2. Moved mosts tests into existing tests.
3. `pd.NA` is are supported only when they are used in a pandas categorical dtype. (This test was added.)

jnothman commented on sklearn/utils/tests/test_encode.py:
```diff
@@ -193,3 +193,12 @@ def test_check_unknown_with_both_missing_values():
+    values = np.array([np.nan, None, float('nan'), 'a'], dtype=object)
+
+    msg = (r"Encoders supports missing values encoded as np\.nan or None and "
+           r"does not support float\('nan'\)")
```
```suggestion
           r"do not currently support float\('nan'\)")
```

jnothman reviewed (commented):
After a bit more thinking, I don't think the policy of not supporting NaNs that are not the numpy singleton is going to work: pickling and unpickling an np.nan is enough for identity- and set-based operations to fail. It's not worth having an implementation that fails on unpickled data. Sorry! :(

thomasjpfan committed:
DOC Update docstring

thomasjpfan committed:
ENH Adds support for nan dtypes

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
CLN Less diffs

jnothman commented:
scikit-learn.scikit-learn (Linux py36_conda_openblas) failure seems unrelated.

Doc build failure is true positive:
```
Unexpected failing examples:
/home/circleci/project/examples/inspection/plot_permutation_importance.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/sphinx_gallery/gen_gallery.py", line 159, in call_memory
    return 0., func()
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/sphinx_gallery/gen_rst.py", line 466, in __call__
    exec(self.code, self.fake_main.__dict__)
  File "/home/circleci/project/examples/inspection/plot_permutation_importance.py", line 103, in <module>
    print("RF test accuracy: %0.3f" % rf.score(X_test, y_test))
  File "/home/circleci/project/sklearn/utils/metaestimators.py", line 119, in <lambda>
    out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)
  File "/home/circleci/project/sklearn/pipeline.py", line 606, in score
    Xt = transform.transform(Xt)
  File "/home/circleci/project/sklearn/compose/_column_transformer.py", line 604, in transform
    Xs = self._fit_transform(X, None, _transform_one, fitted=True)
  File "/home/circleci/project/sklearn/compose/_column_transformer.py", line 467, in _fit_transform
    self._iter(fitted=fitted, replace_strings=True), 1))
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/joblib/parallel.py", line 1029, in __call__
    if self.dispatch_one_batch(iterator):
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/joblib/parallel.py", line 847, in dispatch_one_batch
    self._dispatch(tasks)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/joblib/parallel.py", line 765, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/joblib/_parallel_backends.py", line 208, in apply_async
    result = ImmediateResult(func)
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/joblib/_parallel_backends.py", line 572, in __init__
    self.results = batch()
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/joblib/parallel.py", line 253, in __call__
    for func, args, kwargs in self.items]
  File "/home/circleci/miniconda/envs/testenv/lib/python3.6/site-packages/joblib/parallel.py", line 253, in <listcomp>
    for func, args, kwargs in self.items]
  File "/home/circleci/project/sklearn/pipeline.py", line 713, in _transform_one
    res = transformer.transform(X)
  File "/home/circleci/project/sklearn/preprocessing/_encoders.py", line 456, in transform
    force_all_finite='allow-nan')
  File "/home/circleci/project/sklearn/preprocessing/_encoders.py", line 154, in _transform
    check_unknown=False)
  File "/home/circleci/project/sklearn/utils/_encode.py", line 155, in _encode
    raise ValueError(f"y contains previously unseen labels: {str(e)}")
ValueError: y contains previously unseen labels: nan

```

thomasjpfan committed:
FIX Fixes check_unknown bug with nans

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
CLN Remove unneeded code

jnothman commented on doc/modules/preprocessing.rst:
```diff
@@ -590,6 +590,26 @@ In the transformed `X`, the first column is the encoding of the feature with
+.. note::
+
+    :class:`OneHotEncoder` only supports missing values encoded as `np.nan`
+    or `None` and does not currently support `float('nan')`.
```
I think this is now outdated

thomasjpfan committed:
DOC Update

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -27,13 +28,85 @@ def _unique(values, *, return_inverse=False):
+    missing_values = [value for value in items if is_missing(value)]
+
+    if not missing_values:
+        return items, []
```
Does this imply that `items` is already a set? It should be documented.

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -27,13 +28,85 @@ def _unique(values, *, return_inverse=False):
+            return value is None or is_scalar_nan(value)
+        return is_scalar_nan(value)
+
+    # Return items without missing items
```
I find this confusing
```suggestion
    # Extract the missing items
```

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -27,13 +28,85 @@ def _unique(values, *, return_inverse=False):
+        return is_scalar_nan(value)
+
+    # Return items without missing items
+    missing_values = [value for value in items if is_missing(value)]
+
+    if not missing_values:
+        return items, []
+
+    # Enforces an order where None always comes first
+    if None in missing_values:
+        if len(missing_values) == 1:
+            output_missing_values = [None]
+        else:
+            output_missing_values = [None, np.nan]
+    else:
+        output_missing_values = [np.nan]
+
+    # create set without the missing values
+    output = set(value for value in items if not is_missing(value))
```
I think set difference still works and is more succinct
```suggestion
    missing_values = {value for value in items if is_missing(value)}
    output = items - missing_values

    # Enforces an order where None always comes first
    if None in missing_values:
        if len(missing_values) == 1:
            output_missing_values = [None]
        else:
            output_missing_values = [None, np.nan]
    elif missing_values:
        output_missing_values = [np.nan]
    else:
        output_missing_values = []
```

thomasjpfan replied:
Set different does not work with `float('nan')`:

```
values = [float('nan'), float('nan'), 'a', 'b', 'b', 'b']
uniques = set(values)
uniques - set([float('nan')])
# {'a', 'b', nan, nan}
```

jnothman replied:
I don't think that's quite accurate. Rather, set difference for NaN's will work by object identity rather than value comparison. So as long as the NaNs you are subtracting are the same objects you have in the other operand, it should work fine.

thomasjpfan replied:
Ah, that makes sense. The PR updated with your suggestion.

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -27,13 +28,85 @@ def _unique(values, *, return_inverse=False):
+
+
+def _split_missing(items, none_is_missing=True):
+    """Split items into a set and a list of missing values. If none_is_missing
```
```suggestion
    """Extract missing values from items
    
    Returns the set of items and a list of missing values. If none_is_missing
```

jnothman reviewed (commented):
Partial review

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
CLN Update split missing

thomasjpfan committed:
DOC Adds comments

thomasjpfan committed:
STY Fix

thomasjpfan committed:
CLN Address comments

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -42,9 +133,7 @@ def _unique_python(values, *, return_inverse):
-        table = {val: i for i, val in enumerate(uniques)}
-        inverse = np.array([table[v] for v in values])
-        return uniques, inverse
+        return uniques, _map_to_integer(values, uniques)
```
`_extract_missing` above will treat None as missing, but I don't think `_map_to_integer` does. Is this problematic? If we are allowing a mixed use of NaN and None then I suspect `_nandict` needs a couple of extra `or key is None`.

thomasjpfan replied:
Since `None` can be used as a key in a normal dictionary. This would treat `None` and `nan` as two different categories. Do you think this behavior is surprising?

thomasjpfan replied:
It does simplify things we error in the situations when there is `nan` and `None`.

thomasjpfan replied:
As of now, I think we can support the weird case of `nan` and `None` together.

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -118,26 +206,66 @@ def _check_unknown(values, known_values, return_mask=False):
+
+        def is_valid(value):
+            value_in_set = value in uniques_set
+            if nan_in_uniques and is_scalar_nan(value):
```
This is confusing. It's performing `if bool(nan_in_uniques) and ...` but `bool(nan_in_uniques) == is_missing_in_uniques` and we return the latter.

I think the logic is not right here.

I also thing we should fast path the case where value_in_set.

See suggestion below

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -118,26 +206,66 @@ def _check_unknown(values, known_values, return_mask=False):
-            return diff
+
+        # ensure that None is at the end
+        none_in_diff = None in diff
+        if none_in_diff:
+            diff.remove(None)
+        diff = list(diff)
+        if none_in_diff:
+            diff.append(None)
```
We've got too many variables and too many conditionals in this function.
```suggestion
        if None in diff:
            diff.remove(None)
            diff = list(diff)
            diff.append(None)
        else:
            diff = list(diff)
```

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -118,26 +206,66 @@ def _check_unknown(values, known_values, return_mask=False):
+
+        nan_in_diff = nan_in_values and not nan_in_uniques
+
+        def is_valid(value):
+            value_in_set = value in uniques_set
+            if nan_in_uniques and is_scalar_nan(value):
+                return is_missing_in_uniques
+            return value_in_set
```
```suggestion
		def is_valid(value):
		    if value in uniques_set:
		        return True
		    return is_missing_in_uniques and is_scalar_nan(value)
```

jnothman commented on sklearn/utils/_encode.py:
```diff
@@ -118,26 +206,66 @@ def _check_unknown(values, known_values, return_mask=False):
+    if values.dtype.kind in 'UO':
+        values_set = set(values)
+        values_set, nan_in_values = _extract_missing(values_set,
+                                                     none_is_missing=False)
```
I'm a bit confused about why we use none_is_missing=False.

thomasjpfan replied:
Set difference works by default with `None`. If `None` was included as missing, we would need a little bit more logic to handle `None`. Currently `nan_in_*` can only include `nan`, which makes it slightly easier to figure out if `nan` is in the diff.

Maybe there is a way to simplify this logic a little...

thomasjpfan replied:
I removed this `none_is_missing` and changed the signature of `_extract_missing` to return a `MissingValues` tuple, which better holds the missing value info.

jnothman commented on sklearn/utils/tests/test_encode.py:
```diff
@@ -53,21 +64,145 @@ def test_encode_with_check_unknown():
+
+def test_unique_util_with_both_missing_values():
+    # test for both types of missing values for object dtype
+    values = np.array([np.nan, 'a', 'c', 'c', None, float('nan'),
```
If we're supporting this, please test the case of only NaN in training, but None present in test, and vice-versa.

thomasjpfan replied:
Added test to `test_one_hot_encoder_specified_categories`

jnothman reviewed (commented):
Moving from the pesky NaNs to the pesky Nones...

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
WIP Adds a MissingValues tuple

thomasjpfan committed:
DOC Adds more comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

jnothman reviewed (approved):
I think I'm done here... Lgtm!

Let's solicit another review...

jnothman commented:
We have ?4 weeks for another review to push this into 0.24 ... volunteers?

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

agramfort commented on doc/modules/preprocessing.rst:
```diff
@@ -590,6 +590,21 @@ In the transformed `X`, the first column is the encoding of the feature with
+    >>> enc.transform(X).toarray()
+    array([[0., 1., 0., 0., 1., 0.],
+           [1., 0., 0., 0., 0., 1.],
+           [0., 0., 1., 1., 0., 0.]])
```
can you add a note here about what happens if None and np.nan are present in the same column?

agramfort replied:
I would also pass explicitly here the handle_unknown parameter. To document it.

thomasjpfan replied:
Updated PR:

1. The user guide now describes what happens when `nan` and `None` are present.
2. The default value of `handle_unknown='error'` is passed explicitly.

agramfort reviewed (commented):
besides LGTM

maybe I would not have done with a new MissingValues class but I understand the argument to enforce typing.

thomasjpfan committed:
DOC Updates user guide to include feature with both nan and None

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

thomasjpfan committed:
DOC Spelling

thomasjpfan committed:
DOC Fixes spacing

agramfort reviewed (approved):
LGTM thx @thomasjpfan !

cmarmo commented:
Two approvals here and the opportunity to close three issues and moving forward with the milestone!
 @thomasjpfan do you mind syncing (just to be sure) and then this could probably be merged?

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into one_hot_encoding_missing_category

amueller commented:
OMG THIS WAS MERGED!!! I'm so happy!

jnothman commented:
Yes, thanks @thomasjpfan for putting in the hard work to make this finally happen!!
